src/Playwright/Core/Waiter.cs

```diff
@@ -117,6 +117,12 @@ internal void RejectImmediately(Exception exception)
         _immediateError = exception;
     }
 
+    internal void RejectOnCancellation(CancellationToken cancellationToken)
+    {
+        var cancelledTask = Task.FromCanceled(cancellationToken);
+        RejectOn(cancelledTask, () => cancelledTask.Dispose());
```

Comment:
This should probably look something like this:

```suggestion
        if (!cancellationToken.CanBeCanceled) return;
        var cts = new TaskCompletionSource<bool>();
        var registration = cancellationToken.Register(state => cts.SetCanceled((CancellationToken)state), cancellationToken);
        RejectOn(cts.Task, () => registration.Dispose());
```

Reply:
Thank you very much. I was fighting with this part. I found some approaches but nothing worked properly.
I am adding tests to cover cancellation scenarios.